Libs/MRML/Widgets/qMRMLScalarsDisplayWidget.cxx

```diff
@@ -0,0 +1,611 @@
+void qMRMLScalarsDisplayWidget::setScalarsColorNode(vtkMRMLNode* colorNode)
+{
+  this->setScalarsColorNode(vtkMRMLColorNode::SafeDownCast(colorNode));
+}
+
+//------------------------------------------------------------------------------
+void qMRMLScalarsDisplayWidget::setScalarsColorNode(vtkMRMLColorNode* colorNode)
+{
+  Q_D(qMRMLScalarsDisplayWidget);
+
+  foreach (vtkMRMLDisplayNode* displayNode, d->CurrentDisplayNodes)
+    {
+    if (displayNode)
+      {
+      displayNode->SetAndObserveColorNodeID(colorNode ? colorNode->GetID() : nullptr);
+      }
+    }
+}
+
+//------------------------------------------------------------------------------
+vtkMRMLColorNode* qMRMLScalarsDisplayWidget::scalarsColorNode()const
+{
+  Q_D(const qMRMLScalarsDisplayWidget);
+  return vtkMRMLColorNode::SafeDownCast(d->ScalarsColorNodeComboBox->currentNode());
+}
+
+// --------------------------------------------------------------------------
+void qMRMLScalarsDisplayWidget::setScalarRangeMode(ControlMode controlMode)
+{
+  Q_D(qMRMLScalarsDisplayWidget);
+
+  bool modified = false;
+  foreach (vtkMRMLDisplayNode* displayNode, d->CurrentDisplayNodes)
+    {
+    int newScalarRangeMode = vtkMRMLDisplayNode::UseManualScalarRange;
+    switch (controlMode)
+      {
+      case Data: newScalarRangeMode = vtkMRMLDisplayNode::UseDataScalarRange; break;
+      case LUT: newScalarRangeMode = vtkMRMLDisplayNode::UseColorNodeScalarRange; break;
+      case DataType: newScalarRangeMode = vtkMRMLDisplayNode::UseDataTypeScalarRange; break;
+      case Manual: newScalarRangeMode = vtkMRMLDisplayNode::UseManualScalarRange; break;
+      case DirectMapping: newScalarRangeMode = vtkMRMLDisplayNode::UseDirectMapping; break;
+      }
+
+    int currentScalarRangeMode = displayNode->GetScalarRangeFlag();
+    if (currentScalarRangeMode != newScalarRangeMode)
+      {
+      displayNode->SetScalarRangeFlag(newScalarRangeMode);
+      modified = true;
+      }
+    }
+  if (modified)
+    {
+    emit this->scalarRangeModeValueChanged(controlMode);
+    }
+}
+
+// --------------------------------------------------------------------------
+void qMRMLScalarsDisplayWidget::setScalarRangeMode(int scalarRangeMode)
+{
+  switch(scalarRangeMode)
+    {
+    case 0:
+      this->setScalarRangeMode(qMRMLScalarsDisplayWidget::Data);
+      break;
+    case 1:
+      this->setScalarRangeMode(qMRMLScalarsDisplayWidget::LUT);
+      break;
+    case 2:
+      this->setScalarRangeMode(qMRMLScalarsDisplayWidget::DataType);
+      break;
+    case 3:
+      this->setScalarRangeMode(qMRMLScalarsDisplayWidget::Manual);
+      break;
+    case 4:
+      this->setScalarRangeMode(qMRMLScalarsDisplayWidget::DirectMapping);
+      break;
+    default:
+      break;
+    }
+}
+
+// --------------------------------------------------------------------------
+qMRMLScalarsDisplayWidget::ControlMode qMRMLScalarsDisplayWidget::scalarRangeMode() const
```

Comment:
It would be nicer to have a single map that converts from combobox-item-index to range-mode and use that everywhere.
It is also quite confusing that `setScalarRangeMode` takes a combobox-item-index but `scalarRangeMode` returns a range-mode enum.
Maybe everything is correct, but at least things should be made more clear with more explicitly named arguments and variables.
Moreover, users of this class would probably expect to be able to use vtkMRMLDisplayNode::UseXYZ enums directly instead of thinking about converting to some internal enums.

Reply:
Makes sense. I did wo things:
- Remove the ControlMode (which had a strange name anyway, and it contained the same options as vtkMRMLDisplayNode::ScalarRangeFlagType but in a slightly different order: Manual is with index 3 in ControlMode but with 0 in the ScalarRangeFlagType), and just use vtkMRMLDisplayNode::ScalarRangeFlagType instead
- Reorder the items in the combobox to correspond to the enum values in vtkMRMLDisplayNode::ScalarRangeFlagType, and made the item 1 the default (the same one: Data scalar range (auto), which I renamed from Active scalar (auto))